machine/linux: Switch to virtiofs by default

cgwalters opened the pull request:
machine/linux: Use memory-backend-memfd by default

This is prep for using virtiofsd; it has no real
impact otherwise.

Signed-off-by: Colin Walters <walters@verbum.org>

---

machine: Also initialize virtiofs mounts

I'm hitting a bug with 9p when trying to transfer large files.
In RHEL at least 9p isn't supported because it's known to have a
lot of design flaws; virtiofsd is the supported and recommended
way to share files between a host and guest.

This patch somewhat hackily sets up parallel virtiofsd mounts
alongside the 9p mounts (just on the host side, but one
can then easily mount in the guest).

This is allowing me to do some side-by-side comparisons
with the exact same machine.

I think what we can do in the short term is make
this a hidden option to enable (both "alongside" and also
fully replacing the 9p code).

Signed-off-by: Colin Walters <walters@verbum.org>

---

```release-note
On Linux, podman machine has switched to using virtiofs (instead of 9p) for mounting host filesystems.  This change will be transparently applied when a machine is stopped and restarted (e.g. "podman machine stop; podman machine start") or reinitializing the machine.  This change is intended to improve performance and reliability.
```

cgwalters commented:
https://gist.github.com/cgwalters/3d22459bae05a89d86b9f0d501d82feb for my notes debugging a 9p issue that spawned this

cgwalters commented:
One architectural issue we have now is that the qemu process from `podman machine start` just happens to run in the context of the user's shell.  Yet, when using virtiofsd suddenly we have new processes that need to lifecycle bind with qemu (i.e. they should die if qemu dies).  That kind of happens implicitly with this...but...IMO it'd be better to consider having `podman machine start` actually wrap everything underneath a `systemd-run --user -u podman-machine`.

Luap99 reviewed (commented):
Practically speaking is there any reason to support both 9p and virtiofs at the same time? Supporting both just increases maintenance burden. I guess the only downside is that we depend on virtiofsd.

rhatdan commented:
We should just move to virtiofsd and drop 9p

cgwalters commented:
> Practically speaking is there any reason to support both 9p and virtiofs at the same time? Supporting both just increases maintenance burden. 

It'd definitely simplify this patch to just hard switch.  Basically I was just trying to be conservative; on the flip side it wasn't really *hard* either to do both. 

I am seeing bits about FreeBSD mentioned, which is not supported by virtiofsd (it uses lots of Linux specific namespacing bits, openat2, etc).  Digging around slightly I see https://wiki.freebsd.org/VirtFS which looks like a not-yet-upstream implementation, although that may be mostly relevant when acting as a guest, not a host.  It looks like... @dfr may be a contact here?

dfr commented:
> > Practically speaking is there any reason to support both 9p and virtiofs at the same time? Supporting both just increases maintenance burden.
> 
> It'd definitely simplify this patch to just hard switch. Basically I was just trying to be conservative; on the flip side it wasn't really _hard_ either to do both.
> 
> I am seeing bits about FreeBSD mentioned, which is not supported by virtiofsd (it uses lots of Linux specific namespacing bits, openat2, etc). Digging around slightly I see https://wiki.freebsd.org/VirtFS which looks like a not-yet-upstream implementation, although that may be mostly relevant when acting as a guest, not a host. It looks like... @dfr may be a contact here?

There will be support for 9p in FreeBSD 15.0 - the implementation is derived from the VirtFS code you reference. It is likely that I will also merge that to FreeBSD 14.2 when that is released later this year.

Last week at [BSDCan](https://wiki.freebsd.org/VirtFS), I heard about a FreeBSD implementation of virtiofs which I believe is compatible with the Linux virtiofs but I don't know the status of that and can't predict when it will make it to a FreeBSD release.

Luap99 commented:
> > > Practically speaking is there any reason to support both 9p and virtiofs at the same time? Supporting both just increases maintenance burden.
> > 
> > 
> > It'd definitely simplify this patch to just hard switch. Basically I was just trying to be conservative; on the flip side it wasn't really _hard_ either to do both.
> > I am seeing bits about FreeBSD mentioned, which is not supported by virtiofsd (it uses lots of Linux specific namespacing bits, openat2, etc). Digging around slightly I see https://wiki.freebsd.org/VirtFS which looks like a not-yet-upstream implementation, although that may be mostly relevant when acting as a guest, not a host. It looks like... @dfr may be a contact here?
> 
> There will be support for 9p in FreeBSD 15.0 - the implementation is derived from the VirtFS code you reference. It is likely that I will also merge that to FreeBSD 14.2 when that is released later this year.
> 
> Last week at [BSDCan](https://wiki.freebsd.org/VirtFS), I heard about a FreeBSD implementation of virtiofs which I believe is compatible with the Linux virtiofs but I don't know the status of that and can't predict when it will make it to a FreeBSD release.

We don't support feebsd guest so the question is can virtiofsd run on a freebsd host? If not hard forcing virtiofs with qemu will break the freebsd (host) support for podman machine but tbh I have no idea if there is anyone using this in the first place???

dfr commented:
> > > > Practically speaking is there any reason to support both 9p and virtiofs at the same time? Supporting both just increases maintenance burden.
> > > 
> > > 
> > > It'd definitely simplify this patch to just hard switch. Basically I was just trying to be conservative; on the flip side it wasn't really _hard_ either to do both.
> > > I am seeing bits about FreeBSD mentioned, which is not supported by virtiofsd (it uses lots of Linux specific namespacing bits, openat2, etc). Digging around slightly I see https://wiki.freebsd.org/VirtFS which looks like a not-yet-upstream implementation, although that may be mostly relevant when acting as a guest, not a host. It looks like... @dfr may be a contact here?
> > 
> > 
> > There will be support for 9p in FreeBSD 15.0 - the implementation is derived from the VirtFS code you reference. It is likely that I will also merge that to FreeBSD 14.2 when that is released later this year.
> > Last week at [BSDCan](https://wiki.freebsd.org/VirtFS), I heard about a FreeBSD implementation of virtiofs which I believe is compatible with the Linux virtiofs but I don't know the status of that and can't predict when it will make it to a FreeBSD release.
> 
> We don't support feebsd guest so the question is can virtiofsd run on a freebsd host? If not hard forcing virtiofs with qemu will break the freebsd (host) support for podman machine but tbh I have no idea if there is anyone using this in the first place???

I have not tried to run virtiofsd on a FreeBSD host and currently 'podman machine' would need some work to support the native hypervisor and that is far down my priority list. The only FreeBSD related thing which is close is #19939 but it doesn't seem likely that will land any time soon and that is related to guest, not host.

It is clear that virtiofs is the way forward and that 9p will be phased out at some point. Personally I would prefer that happen later rather than sooner given that 9p works well on FreeBSD and will be part of an upcoming release whereas virtiofs seems further off. I don't know how to estimate the negative impact of maintaining both in the meantime though.

cgwalters commented:
> I have not tried to run virtiofsd on a FreeBSD host and currently 'podman machine' would need some work to support the native hypervisor and that is far down my priority list.

Does that mean podman machine doesn't work at all today on FreeBSD, and the support that exists (sorry I am just discovering this stuff now) is about running containers directly on the FreeBSD host?  Does that actually work?  Ahhh....I see https://www.freshports.org/sysutils/ocijail/

> It is clear that virtiofs is the way forward and that 9p will be phased out at some point. Personally I would prefer that happen later rather than sooner given that 9p works well on FreeBSD and will be part of an upcoming release whereas virtiofs seems further off. I 

To be clear this is only about "podman machine" which is currently obscure even on Linux (that said, we're forcing its usage as part of some bootc workflows).  And I suspect podman on FreeBSD is already obscure, so we're talking about an obscure corner of an obscure corner...

Here's my proposal: I'll structure things as two patches:

- Add virtiofsd support (as it is now)
- A followup patch which deletes 9p

Then anyone who cares could try to argue for a revert of the second patch later, or at least use it as a starting point.

Luap99 commented:
> Then anyone who cares could try to argue for a revert of the second patch later, or at least use it as a starting point.

SGTM

dfr commented:
> > Then anyone who cares could try to argue for a revert of the second patch later, or at least use it as a starting point.
> 
> SGTM

That works for me. My only use of 'podman machine' is plain vanilla Linux guest on macos host. For anything else, I have FreeBSD hosts and VMs as well as Linux VMs.

cgwalters commented:
OK I cleaned this up a bit more, and split it into the promised two separate patches, one to add virtiofs in parallel, and one to switch to it by default and drop 9p.

In digging through this, a few other notes:

- We were serializing the `Type` of the mount into the machine JSON; I changed the code to just ignore that.  IMO we should serialize less stuff into this JSON and make it dynamic (xref https://github.com/containers/podman/pull/21192 e.g.).
- There's this "client9p" stuff that I *think* is only used on Hyper-V/Windows, is that right?  I didn't touch it and I *think* it won't be affected by this, it had parallel 9p logic.

cgwalters commented on pkg/machine/qemu/stubber.go:
```diff
@@ -95,6 +104,26 @@ func (q *QEMUStubber) setQEMUCommandLine(mc *vmconfigs.MachineConfig) error {
+	// This sets things up so that the `.` we passed in the arguments is the target directory
+	cmd.Dir = directory
+	// Quiet the daemon by default
+	cmd.Env = append(cmd.Env, "RUST_LOG=ERROR")
```
Good catch!

cgwalters commented on pkg/machine/qemu/stubber.go:
```diff
@@ -141,6 +170,53 @@ func runStartVMCommand(cmd *exec.Cmd) error {
+				break
+			}
+			logrus.Debugf("waiting for virtiofsd socket %q", virtiofsCharPath.Path)
+			time.Sleep(time.Second)
```
Fixed by dropping to 100ms, though a slightly better fix would be to use inotify.

That said, IMO the most clean fix (per above) would actually be switching to using one or more systemd units for the qemu process and our helpers; we could have systemd allocate the socket and pass it to virtiofsd or so.  Or failing that, virtiofsd can signal its readiness via the systemd protocol, etc.

cgwalters commented on pkg/machine/qemu/virtiofsd.go:
```diff
@@ -0,0 +1,107 @@
+)
+
+// We should probably make these bulid time configurable too
+const virtiofsdPathFedora = "/usr/libexec/virtiofsd"
+const virtiofsdPathDebian = "/usr/lib/qemu/virtiofsd"
```
Ah man, and now that I look Arch Linux puts it in [`/usr/lib/virtiofsd`](https://archlinux.org/packages/extra/x86_64/virtiofsd).  Fun.  I guess we probably should make this build-time configuration.

Luap99 replied:
Maybe we should use config.FindHelperBinary(), that looks in /usr/libexec/podman and some other places.
I think we could then just do the podman packages do a symlink to /usr/libexec/virtiofsd on fedora and the other distros could create the correct symlink for their packages.

cgwalters replied:
Good idea!  That makes sense, it should work like `catatonit` for example.  Done.

baude commented:
if I am using podman machine on Linux now and you switch to virtiofs, what happens to my machine?  the update process would allow you to update a client and NOT have to recreate the machine.  I don't see how you can remove 9p without that consideration and I also don't think migration is worth it.  As for switching to virtiofs as a default, we usually do not add the function and then immediately make it the default.  we usually have at least one release off bake time.  I could be convinced otherwise.

cgwalters commented:
> if I am using podman machine on Linux now and you switch to virtiofs, what happens to my machine? the update process would allow you to update a client and NOT have to recreate the machine.

The semantics I believe are easy to describe: When podman is updated to a binary that includes this PR, existing machines stay unchanged until you stop/restart them.

>  I don't see how you can remove 9p without that consideration and I also don't think migration is worth it.

There's not much a "migration" per above, it's just a stop/restart because the mounts are done dynamically over SSH.

> As for switching to virtiofs as a default, we usually do not add the function and then immediately make it the default. we usually have at least one release off bake time. I could be convinced otherwise.

Well, that's a debate between you and Dan and Paul right?  I originally did the PR making it opt in, and it's easy to just drop the top commit here.  I could also do a small variant where we make virtiofs the default with an env fallback to 9p.

But the argument here for a default switch IMO is:

- virtiofs is the production supported path for this in RHEL, and while 9p is enabled in Fedora IMO people only use it out of inertia mostly.
- virtiofs is used on MacOS
- The set of podman machine users on Linux is relatively small anyways (although notably again bootc use cases are pushing that up) but I think we can just do some extra "real world", manual testing on this before the next release, right?  I can say it works well for me whereas 9p was buggy.

baude commented:
Well, I did say I could be convinced, but it is a departure from the norm; and that is notable.  How this applies to RHEL should probably be discussed in further detail, because there are subtleties on Podman machine support there.

cgwalters commented:
Updates:

- I discovered and learned how to run `make localmachine`, and that found an important bug that `:Z` didn't work
- Then I noticed that the MacOS+virtiofs code had hardcoded logic to set the security context (and has a totally different flow to set up the mounts via ignition?) - anyways fixed both to use a shared context


But then suddenly `make localmachine` started failing for me with:

<details>

```
run basic podman commands
/var/home/walters/src/github/containers/podman/pkg/machine/e2e/basic_test.go:20
  Podman ops with port forwarding and gvproxy
  /var/home/walters/src/github/containers/podman/pkg/machine/e2e/basic_test.go:104
  > Enter [BeforeEach] run basic podman commands - /var/home/walters/src/github/containers/podman/pkg/machine/e2e/basic_test.go:26 @ 06/06/24 16:39:38.952
--> copying "/tmp/podman-machine-daily.x86_64.qemu.qcow2" to "/tmp/podman_test2944015414/podman-machine-daily.x86_64.qemu.qcow2"
  < Exit [BeforeEach] run basic podman commands - /var/home/walters/src/github/containers/podman/pkg/machine/e2e/basic_test.go:26 @ 06/06/24 16:39:39.385 (433ms)
  > Enter [It] Podman ops with port forwarding and gvproxy - /var/home/walters/src/github/containers/podman/pkg/machine/e2e/basic_test.go:104 @ 06/06/24 16:39:39.385
  /var/home/walters/src/github/containers/podman/bin/podman machine init --disk-size 11 --image /tmp/podman_test2944015414/podman-machine-daily.x86_64.qemu.qcow2 --now 9f8a445b3db3
  Image resized.
  Machine init complete
  Starting machine "9f8a445b3db3"

  This machine is currently configured in rootless mode. If your containers
  require root permissions (e.g. ports < 1024), or if you run into compatibility
  issues with non-podman clients, you can switch using the following command:

        podman machine set --rootful 9f8a445b3db3

  Mounting volume... /tmp/podman_test2944015414:/tmp/podman_test2944015414
  API forwarding listening on: /tmp/podman_test2944015414/podman/9f8a445b3db3-api.sock
  You can connect Docker API clients by setting DOCKER_HOST using the
  following command in your terminal session:

          export DOCKER_HOST='unix:///tmp/podman_test2944015414/podman/9f8a445b3db3-api.sock'

  Machine "9f8a445b3db3" started successfully
  /var/home/walters/src/github/containers/podman/bin/podman -r run -dt --name test -p 62544:80 quay.io/libpod/alpine_nginx
  Error: unknown shorthand flag: 'r' in -r
  See 'podman --help'
  [FAILED] Expected
      <int>: 125
  to match exit code:
      <int>: 0
  In [It] at: /var/home/walters/src/github/containers/podman/pkg/machine/e2e/basic_test.go:115 @ 06/06/24 16:39:51.274

  Full Stack Trace
    github.com/containers/podman/v5/pkg/machine/e2e_test.init.func2.5()
        /var/home/walters/src/github/containers/podman/pkg/machine/e2e/basic_test.go:115 +0x3b4
  < Exit [It] Podman ops with port forwarding and gvproxy - /var/home/walters/src/github/containers/podman/pkg/machine/e2e/basic_test.go:104 @ 06/06/24 16:39:51.274 (11.888s)
  > Enter [AfterEach] run basic podman commands - /var/home/walters/src/github/containers/podman/pkg/machine/e2e/basic_test.go:29 @ 06/06/24 16:39:51.274
  /var/home/walters/src/github/containers/podman/bin/podman machine rm --force 9f8a445b3db3
  Waiting for VM to exit...
  < Exit [AfterEach] run basic podman commands - /var/home/walters/src/github/containers/podman/pkg/machine/e2e/basic_test.go:29 @ 06/06/24 16:39:52.465 (1.191s)
• [FAILED] [13.512 seconds]
run basic podman commands
/var/home/walters/src/github/containers/podman/pkg/machine/e2e/basic_test.go:20
  [It] Podman ops with port forwarding and gvproxy
  /var/home/walters/src/github/containers/podman/pkg/machine/e2e/basic_test.go:104
```
<details>
which I'm debugging now...

rhatdan commented:
Since almost no one ever changes the default, it is better to default to virtiofs. Falling back to plan9. 
virtiofsd performs better, is better supported (the engineers who wrote it work for Red Hat).  And we use it on Mac.

Luap99 commented:
I am in favour of directly switching to virtiofs because it is faster and works better so I really see no reason to keep 9p around and pay the cost of maintaining both. Because if we do both it means we a) have to tests both code paths  and b) when getting bug reports about it have to know which mounts a user is actually using. By doing a hard switch we just test one thing and I can tell by the version what will be used.

For me the only downside is that what now have a hard dependency on virtiofsd for podman-machine. So we need to ensure the few podman machine users on linux are aware of it. I think qemu isn't a hard dependency in package because we really do not need for containers so virtiofsd would not be different.

Luap99 commented on pkg/machine/qemu/virtiofsd.go:
```diff
@@ -0,0 +1,116 @@
+		}
+	}
+	if binaryPath == "" {
+		return nil, fmt.Errorf("failed to find virtiofsd")
```
please return the original error from FindHelperBinary because that actually tells the user how to configure it in case virtiofsd is installed.

Luap99 commented on pkg/machine/qemu/virtiofsd.go:
```diff
@@ -0,0 +1,116 @@
+	}
+	binaryPath, err = cfg.FindHelperBinary("virtiofsd", true)
+	if err != nil {
+		logrus.Debugf("Failed to find virtiofsd in helper dir, using fallback")
+		for _, p := range []string{virtiofsdPathDebian, virtiofsdPathFedora, virtiofsdPathArch} {
+			if err := fileutils.Exists(p); err == nil {
+				binaryPath = p
+				break
+			}
+		}
```
Do we really want to do that? If we allow these paths than any distro just wants to be added here instead of configuring it once. IMO we shouldn't try these at all and require the symlink setup or require this to be set at build time and/or patched.

cgwalters replied:
> IMO we shouldn't try these at all and require the symlink setup or require this to be set at build time and/or patched.

Fair, but I think there's going to be some short term pain from that in terms of things like running podman machine from git not working unless you've got the symlink on your host's `/usr/libexec/podman` setup.  But yes, done.

cgwalters commented:
> For me the only downside is that what now have a hard dependency on virtiofsd for podman-machine. So we need to ensure the few podman machine users on linux are aware of it. I think qemu isn't a hard dependency in package because we really do not need for containers so virtiofsd would not be different.

Right.  On this topic it seems like the CI environment in cirrus needs to install virtiofsd...I think I may have just found the right place for that...will try.

cgwalters commented on pkg/machine/qemu/stubber.go:
```diff
@@ -168,6 +166,28 @@ func (q *QEMUStubber) StartVM(mc *vmconfigs.MachineConfig) (func() error, func()
+			q.virtiofsHelpers = append(q.virtiofsHelpers, *virtiofsdHelper)
+		}
+	} else {
+		panic("Unhandled mount type:" + q.mountType)
```
In Rust this would have started out as
```
enum Virtfs {
  NineP,
  Virtiofs
}
```

And this bit of code would have been a compiler-checked exhaustive `match` that went down to one element.  But there's no such thing in Go...changing this to return an error would be weird since it should be really unreachable code, rihgt?

Probably what I can do though is just eliminate the mountType variable in the last patch that drops 9p.

cgwalters commented on pkg/machine/qemu/stubber.go:
```diff
@@ -168,6 +166,28 @@ func (q *QEMUStubber) StartVM(mc *vmconfigs.MachineConfig) (func() error, func()
+				return nil, nil, fmt.Errorf("failed to init virtiofsd for mount %s: %w", hostmnt.Source, err)
+			}
+			q.Command = append(q.Command, qemuArgs...)
+			q.virtiofsHelpers = append(q.virtiofsHelpers, *virtiofsdHelper)
```
Ah yes, we were returning a "readyfunc" which released the qemu command, I changed it to cleanup the virtiofsd helpers too.

cgwalters commented:
OK cool, my semi-educated guess for the right place to install virtiofsd for the CI runs worked.  The mac/windows machine tests failed on the new test I think because we were only setting up the volume on qemu but in the interest of expediency I'm changing that test to just test linux.

cgwalters commented:
> Well, I did say I could be convinced, but it is a departure from the norm; 

By "norm" do you more mean "status quo for podman-machine", or are there other prominent uses of 9p outside of this codebase that you consider "normal"?  I think for a lot of people this is mainly because 9p happens to be built into qemu, whereas virtiofs is implemented externally.  But other virt systems, such as crosvm ([ref](https://crosvm.dev/book/devices/fs.html)) and cloud-hypervisor ([ref](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/81b41b55fda136385bc94a8417227a50277fff2b/docs/fs.md?plain=1#L1)) AFAIK only support virtiofs.

> and that is notable. How this applies to RHEL should probably be discussed in further detail, because there are subtleties on Podman machine support there.

My reference to RHEL was primarily from the angle of "virtiofs is in the set of things more broadly tested/supported with respect to virtualization" and not about using podman machine with a RHEL host really.  Though it is notable this patch would be required for any effort to use a RHEL system as the guest.

cgwalters commented:
OK, there was a podman-machine+windows timeout, but it seemed to have been a flake (and as far as I know there's no real intersection here with the Windows codepaths).  Restarting fixed it.  I added a release note.  So I think this one should be generally ready.

rhatdan commented on pkg/machine/apple/apple.go:
```diff
@@ -87,7 +87,7 @@ func GenerateSystemDFilesForVirtiofsMounts(mounts []machine.VirtIoFs) ([]ignitio
 		mountUnit.Add("Mount", "Where", "%s")
 		mountUnit.Add("Mount", "Type", "virtiofs")
-		mountUnit.Add("Mount", "Options", "context=\"system_u:object_r:nfs_t:s0\"")
+		mountUnit.Add("Mount", "Options", fmt.Sprintf("context=\"%s\"", machine.NFSSELinuxContext))
```
Nit use %q

cgwalters replied:
Sure, will do; just waiting to see if anyone else has other comments before doing another update and round of CI.

rhatdan commented:
LGTM
@Luap99 PTAL

cgwalters commented on contrib/cirrus/setup_environment.sh:
```diff
@@ -410,7 +410,12 @@ case "$TEST_FLAVOR" in
     machine-linux)
-        showrun dnf install -y podman-gvproxy*
+        showrun dnf install -y podman-gvproxy* virtiofsd
+        # Bootstrap this link if it isn't yet in the package; xref
```
I was thinking about this a bit more, and one thing that will probably show up is static package analyzers will warn about the broken symlink (presumably we won't take a hard dep on virtiofsd).

I don't know how annoying it will be though.

Luap99 replied:
Yeah I do wonder if we should add a new podman-machine package that just contains the necessary deps as requires and could add the symlink there. That way users don't get broken every time we add a new dep and they can just have the one mostly empty package just to get further packages.

cgwalters replied:
Yeah, makes sense.

Luap99 commented:
Also note that I still consider @baude to have the final say on this.

germag commented:
Moving to virtiofsd, while running it inside a user namespace (inside a `podman unshare` or using its own sandbox `--uid-map/--gid-map`) will allow sharing the user image storage and mount it inside the VM in `/usr/lib/containers/storage`. So, allowing the root to run containers using the user's images without copy them or requiring to store the image in the (VM)root storage. 

This is important for `podman-bootc` since we need to run as root to create disk images, and currently we force linux user to store the images in  the (VM)root user.

germag commented:
Btw, regarding FreeBSD, we plan to support virtiofsd on it in the near future, @stefano-garzarella is doing all the required modification on rust-vmm

baude commented:
/approve

cgwalters commented:
Thanks for the lgtm, Brent !

One thing I'd re-emphasize a bit here is two things:

- Currently "podman machine" is critical path for my day to day work on Linux too
- You know I'm going to be around to help with any fallout from this PR as well as ongoing maintenance

cgwalters commented:
Can we queue this for a 5.1 release? Getting more reports, e.g. https://github.com/osbuild/bootc-image-builder/issues/540

Luap99 commented:
This isn't the kind of stuff that should go into a patch release given it requires packaging changes IMO.
Regardless 5.2 will be released in the next few weeks anyway

cgwalters commented:
> Regardless 5.2 will be released in the next few weeks anyway

Is that planned to be released to fedora 40 for example?

Luap99 commented:
Yes

rhatdan commented:
All minor releases are released to supported Fedora's.
